src/editors/EditorPageProvider.jsx

```diff
@@ -0,0 +1,93 @@
+import React, {
+  useState, useEffect, useRef, useMemo,
+} from 'react';
+import PropTypes from 'prop-types';
+import { fetchBlockById, fetchUnitById, saveBlock } from './data/api';
+import EditorPageContext from './EditorPageContext';
+import { ActionStates } from './data/constants';
+
+/* This Component serves as a container for state for V2 editors,
+to avoid prop drilling for: saving, loading, and navigating away from content. */
+
+const EditorPageProvider = ({
+  blockType, courseId, blockId, studioEndpointUrl, children,
+}) => {
+  const editorRef = useRef(null);
+  const [blockValue, setBlockValue] = useState(null); // this is the intial block, as called in from the api.
+  const [blockError, setBlockError] = useState(null);
+  const [blockLoading, setBlockLoading] = useState(ActionStates.NOT_BEGUN);
+  const [unitUrl, setUnitUrlValue] = useState(null);
+  const [unitUrlError, setUnitUrlError] = useState(null);
+  const [unitUrlLoading, setUnitUrlLoading] = useState(ActionStates.NOT_BEGUN);
+  const [blockContent, setBlockContent] = useState(null); // This is the updated content to be saved via api call
+  const [saveResponse, setSaveResponse] = useState(null);
+  const [saveUnderway, setSaveUnderway] = useState(ActionStates.NOT_BEGUN);
+
+  const value = useMemo(() => ({
+    editorRef,
+    blockValue,
+    blockError,
+    blockLoading,
+    unitUrl,
+    unitUrlError,
+    unitUrlLoading,
+    setBlockContent,
+    saveResponse,
+    setSaveUnderway,
+    saveUnderway,
+    studioEndpointUrl,
+    blockId,
+    courseId,
+    blockType,
+  }), [blockLoading, unitUrlLoading, saveUnderway]);
+
+  useEffect(() => {
```

Comment:
So we want to run "useeffect" as little as possible. We only want it to run:
1. when we initialize, to start the fetching of data
2. When the save process has been begun. 
The second argument to useEffect is an array of dependancies, which trigger a call to useEffect async when they are updated. So, we set our dependency to be `saveUnderway` as it has the following lifecycle:
a. On page load, it is set to value `not_begun`, triggering the fetches. (solving for 1.)
b. On the click of the save button (in footer) saveUnderway = `in_progress` which satifes 2.

If we added more dependancies, it would run every time they updated, causing more re-renders of the all the children of the provider. Does that make sense?

Reply:
Should I add a comment detailing this choice?